### PR TITLE
#361: added an extract placeholder, and a placeholder test-case

### PR DESCRIPTION
**Pull request by MattPerron**
This PR also adds typechecking to function arguments.

### CODE REVIEW

**Comment by apavlo** [comment]
@jarulraj is going to review this now.

**Comment by malin1993ml** [comment]
@jarulraj @MattPerron I'm wondering what's the status on this PR?

**Comment by MattPerron** [comment]
@malin1993ml I'm still waiting for @jarulraj to review

**Comment by apavlo** [comment]
Yeah what gives?!?

**Comment by MattPerron** [comment]
Reviewing this PR is definitely taking a backseat to dates.

**Comment by apavlo** [comment]
@jarulraj *cough*

**Comment by jarulraj** [comment]
Why does Github keep changing the review system -- I apparently forgot to click the "submit review" button :)

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -523,30 +516,28 @@ std::unique_ptr<storage::DataTable> Catalog::CreateMetricsCatalog(
...
-      type::Type::INTEGER, type::Type::GetTypeSize(type::Type::INTEGER),
-      "database_id", true);
+  auto database_id_column = catalog::Column(type::Type::INTEGER,
+      type::Type::GetTypeSize(type::Type::INTEGER), "database_id", true);
   database_id_column.AddConstraint(
       catalog::Constraint(CONSTRAINT_TYPE_NOTNULL, not_null_constraint_name));
 
-  auto database_name_column = catalog::Column(
-      type::Type::VARCHAR, max_name_size, "database_name", true);
+  auto database_name_column = catalog::Column(type::Type::VARCHAR,
```
Comment (MattPerron) [inline src/catalog/catalog.cpp]: I actually didn't touch this part, these changes are due to formatting changes only

File: test/expression/expression_test.cpp [hunk 2]
```diff
@@ -33,22 +31,52 @@ namespace test {
...
+
+  // these need unique ptrs to clean them
+  auto extract_expr = ExpPtr(new expression::FunctionExpression("extract", { part, timestamp }));
+
+  expression::ExpressionUtil::TransformExpression(nullptr, extract_expr.get());
+
+
+  // do a lookup (we pass null schema because there are no tuple value expressions
+  EXPECT_TRUE(
+      extract_expr->Evaluate(nullptr, nullptr, nullptr).IsNull());
```
Comment (MattPerron) [inline test/expression/expression_test.cpp]: Currently this only happens in tests, So i don't think we should add it as a function.

File: src/catalog/catalog.cpp [hunk 3]
```diff
@@ -751,15 +744,16 @@ Catalog::~Catalog() {
 }
 
 // add amd get methods for UDFs
-void Catalog::AddFunction(
-    const std::string &name, const size_t num_arguments,
+void Catalog::AddFunction(const std::string &name,
```
Comment (MattPerron) [inline src/catalog/catalog.cpp]: This is the only place I touched the catalog.

File: src/include/expression/function_expression.h [hunk 4]
```diff
@@ -23,69 +23,99 @@ namespace expression {
...
+          "Unexpected number of arguments to function: " + func_name_
+              + ". Expected: " + std::to_string(arg_types.size()) + " Actual: "
+              + std::to_string(children_.size()));
     }
-    for (auto &child : children){
-      children_.push_back(std::unique_ptr<AbstractExpression>(child));
+    // check that the types are correct
+    for (size_t i = 0; i < arg_types.size(); i++) {
+      if (children[i]->GetValueType() != arg_types[i]) {
+        throw Exception(
```
Comment (MattPerron) [inline src/include/expression/function_expression.h]: I changed this to throw an ExpressionException

File: src/include/expression/function_expression.h [hunk 5]
```diff
@@ -23,69 +23,99 @@ namespace expression {
...
     return_value_type_ = val_type;
-    if (num_args != children_.size()){
-      throw Exception("Unexpected number of arguments to function: "+func_name_+". Expected: "+std::to_string(num_args)+ " Actual: " + std::to_string(children_.size()));
+    if (arg_types.size() != children_.size()) {
+      throw Exception(
+          "Unexpected number of arguments to function: " + func_name_
+              + ". Expected: " + std::to_string(arg_types.size()) + " Actual: "
+              + std::to_string(children_.size()));
+    }
+    // check that the types are correct
```
Comment (MattPerron) [inline src/include/expression/function_expression.h]: Done

**Comment by MattPerron** [comment]
I made the requested changes, so this should be ready to merge as long as Joy is okay with it

File: test/expression/expression_test.cpp (hunk 2 above)
Comment (jarulraj) [inline test/expression/expression_test.cpp]: Sounds good.

File: src/catalog/catalog.cpp [hunk 6]
```diff
@@ -67,8 +68,8 @@ void Catalog::CreateMetricsCatalog() {
 void Catalog::CreateCatalogDatabase() {
   storage::Database *database = new storage::Database(START_OID);
```
Comment (jarulraj) [inline src/catalog/catalog.cpp]: Oh ok, but it would be great if you could refactor it now.

File: src/catalog/catalog.cpp (hunk 1 above)
Comment (jarulraj) [inline src/catalog/catalog.cpp]: Ok, let's skip this for now.

**Comment by jarulraj** [comment]
Sure, this is good to go if you can take care of the first suggestion.

File: src/catalog/catalog.cpp (hunk 6 above)
Comment (MattPerron) [inline src/catalog/catalog.cpp]: I don't think we should do this now because I think we are planning to do a catalog rewrite next semester anyway.

Comment (jarulraj) [inline src/catalog/catalog.cpp]: Ok, let's then skip this one as well.

**Pull request merged**